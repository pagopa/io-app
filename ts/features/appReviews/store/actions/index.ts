import { ActionType, createStandardAction } from "typesafe-actions";
import { AppFeedbackUri } from "../../../../../definitions/content/AppFeedbackUri";

export type TopicKeys = keyof AppFeedbackUri;

export const appReviewPositiveFeedback = createStandardAction(
  "APP_REVIEW_POSITIVE_FEEDBACK"
)();

export const appReviewNegativeFeedback = createStandardAction(
  "APP_REVIEW_NEGATIVE_FEEDBACK"
)<TopicKeys>();

// Reset feedback data just for testing purposes
export const clearFeedbackDatas = createStandardAction(
  "APP_REVIEW_RESET_FEEDBACK"
)();

export type AppFeedbackActions =
  | ActionType<typeof appReviewPositiveFeedback>
  | ActionType<typeof appReviewNegativeFeedback>
  | ActionType<typeof clearFeedbackDatas>;
