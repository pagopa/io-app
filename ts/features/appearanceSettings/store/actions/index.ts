/**
 * Action types and action creator related to the Appearance settings.
 */
import { ActionType, createStandardAction } from "typesafe-actions";

export const setShowAppearanceSettingsBanner = createStandardAction(
  "SHOW_APPEARANCE_SETTINGS_BANNER"
)<boolean>();

export type AppearanceSettingsActions = ActionType<
  typeof setShowAppearanceSettingsBanner
>;
