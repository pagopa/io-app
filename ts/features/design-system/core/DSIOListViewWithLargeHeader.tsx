import { ListItemHeader, ListItemInfo } from "@pagopa/io-app-design-system";
import { useState } from "react";
import { Alert } from "react-native";
import { IOListViewWithLargeHeader } from "../../../components/ui/IOListViewWithLargeHeader";
import I18n from "../../../i18n";

export const DSIOListViewWithLargeHeader = () => {
  const [refreshing, setRefreshing] = useState(false);

  const pullAction = () => {
    setRefreshing(true);
    Alert.alert("Alert", "Action triggered");
    setTimeout(() => {
      setRefreshing(false);
    }, 2000);
  };

  const renderItems: Array<ListItemInfo & { id: string }> = [
    {
      id: "1",
      label: I18n.t("authentication.cie.nfc.listItemLabel1"),
      value: I18n.t("authentication.cie.nfc.listItemValue1"),
      icon: "systemSettingsAndroid"
    },
    {
      id: "2",
      label: I18n.t("authentication.cie.nfc.listItemLabel2"),
      value: I18n.t("authentication.cie.nfc.listItemValue2"),
      icon: "systemAppsAndroid"
    },
    {
      id: "3",
      label: I18n.t("authentication.cie.nfc.listItemLabel3"),
      value: I18n.t("authentication.cie.nfc.listItemValue3"),
      icon: "systemToggleInstructions"
    },
    {
      id: "4",
      label: I18n.t("authentication.cie.nfc.listItemLabel3"),
      value: I18n.t("authentication.cie.nfc.listItemValue3"),
      icon: "systemToggleInstructions"
    },
    {
      id: "5",
      label: I18n.t("authentication.cie.nfc.listItemLabel3"),
      value: I18n.t("authentication.cie.nfc.listItemValue3"),
      icon: "systemToggleInstructions"
    },
    {
      id: "6",
      label: I18n.t("authentication.cie.nfc.listItemLabel3"),
      value: I18n.t("authentication.cie.nfc.listItemValue3"),
      icon: "systemToggleInstructions"
    },
    {
      id: "7",
      label: I18n.t("authentication.cie.nfc.listItemLabel3"),
      value: I18n.t("authentication.cie.nfc.listItemValue3"),
      icon: "systemToggleInstructions"
    },
    {
      id: "8",
      label: I18n.t("authentication.cie.nfc.listItemLabel3"),
      value: I18n.t("authentication.cie.nfc.listItemValue3"),
      icon: "systemToggleInstructions"
    },
    {
      id: "9",
      label: I18n.t("authentication.cie.nfc.listItemLabel1"),
      value: I18n.t("authentication.cie.nfc.listItemValue1"),
      icon: "systemSettingsAndroid"
    },
    {
      id: "10",
      label: I18n.t("authentication.cie.nfc.listItemLabel2"),
      value: I18n.t("authentication.cie.nfc.listItemValue2"),
      icon: "systemAppsAndroid"
    }
  ];

  return (
    <IOListViewWithLargeHeader
      data={renderItems}
      title={{
        label: I18n.t("authentication.cie.nfc.title")
      }}
      keyExtractor={item => item.id}
      subtitle={I18n.t("authentication.cie.nfc.subtitle")}
      ListHeaderComponent={
        <ListItemHeader
          label={I18n.t("authentication.cie.nfc.listItemTitle")}
        />
      }
      headerActionsProp={{
        showHelp: true
      }}
      renderItem={({ item }) => (
        <ListItemInfo
          {...item}
          accessibilityLabel={`${item.label}; ${item.value}`}
        />
      )}
      refreshControlProps={{
        onRefresh: pullAction,
        refreshing
      }}
    />
  );
};
