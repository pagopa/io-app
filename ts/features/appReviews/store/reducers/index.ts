import { getType } from "typesafe-actions";
import { PersistConfig, persistReducer } from "redux-persist";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { Action } from "../../../../store/actions/types";
import {
  appReviewNegativeFeedback,
  appReviewPositiveFeedback,
  clearFeedbackDatas,
  TopicKeys
} from "../actions";
import {
  logoutFailure,
  logoutSuccess
} from "../../../../store/actions/authentication";

// String formats for the dates represents the log to be persisted to check the last feedback request time
// both fields are updated directly from reducer on action dispatch and not from the action payload.
export type AppFeedbackState = {
  positiveFeedbackDate: string;
  negativeFeedbackDate:
    | Record<TopicKeys, string | undefined>
    | Record<string, never>;
};

export const appFeedbackInitialState: AppFeedbackState = {
  positiveFeedbackDate: "",
  negativeFeedbackDate: {}
};

const appFeedbackReducer = (
  state: AppFeedbackState = appFeedbackInitialState,
  action: Action
): AppFeedbackState => {
  switch (action.type) {
    case getType(logoutSuccess):
    case getType(logoutFailure):
    case getType(clearFeedbackDatas):
      return {
        ...state,
        ...appFeedbackInitialState
      };
    case getType(appReviewPositiveFeedback):
      return {
        ...state,
        positiveFeedbackDate: new Date().toISOString()
      };
    case getType(appReviewNegativeFeedback):
      return {
        ...state,
        negativeFeedbackDate: {
          ...state.negativeFeedbackDate,
          [action.payload]: new Date().toISOString()
        } as Record<TopicKeys, string | undefined>
      };
    default:
      return state;
  }
};

export const persistConfig: PersistConfig = {
  key: "appFeedback",
  storage: AsyncStorage
};

export const appFeedbackPersistor = persistReducer<AppFeedbackState, Action>(
  persistConfig,
  appFeedbackReducer
);
