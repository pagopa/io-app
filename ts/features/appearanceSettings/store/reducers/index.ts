import AsyncStorage from "@react-native-async-storage/async-storage";
import { persistReducer, PersistConfig } from "redux-persist";
import { getType } from "typesafe-actions";
import { differentProfileLoggedIn } from "../../../../store/actions/crossSessions";
import { Action } from "../../../../store/actions/types";
import { setShowAppearanceSettingsBanner } from "../actions";

export type AppearanceSettingsState = {
  showAppearanceBanner: boolean;
};

export const appearanceSettingsReducerInitialState = {
  showAppearanceBanner: true
};

const appearanceSettingsReducer = (
  state: AppearanceSettingsState = appearanceSettingsReducerInitialState,
  action: Action
): AppearanceSettingsState => {
  switch (action.type) {
    case getType(setShowAppearanceSettingsBanner): {
      return {
        ...state,
        showAppearanceBanner: action.payload
      };
    }
    case getType(differentProfileLoggedIn): {
      return appearanceSettingsReducerInitialState;
    }
    default:
      return state;
  }
};

const persistConfig: PersistConfig = {
  key: "appearanceSettings",
  storage: AsyncStorage
};

export const appearanceSettingsReducerPersistor = persistReducer(
  persistConfig,
  appearanceSettingsReducer
);
