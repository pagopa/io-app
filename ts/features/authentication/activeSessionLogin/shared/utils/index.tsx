import I18n from "i18next";
import { Millisecond } from "@pagopa/ts-commons/lib/units";
import { Platform } from "react-native";
import { CieAuthenticationErrorReason } from "../../../login/cie/store/actions/cie";
import { ReadingState } from "../../../login/cie/components/CieCardReadingAnimation";

// the timeout we sleep until move to consent form screen when authentication goes well
export const WAIT_TIMEOUT_NAVIGATION = 1700 as Millisecond;
export const WAIT_TIMEOUT_NAVIGATION_ACCESSIBILITY = 5000 as Millisecond;
export const VIBRATION = 100 as Millisecond;
export const accessibityTimeout = 100 as Millisecond;

// A subset of Cie Events (errors) which is of interest to analytics
export const analyticActions = new Map<CieAuthenticationErrorReason, string>([
  // Reading interrupted before the sdk complete the reading
  ["Transmission Error", I18n.t("authentication.cie.card.error.onTagLost")],
  ["ON_TAG_LOST", I18n.t("authentication.cie.card.error.onTagLost")],
  [
    "TAG_ERROR_NFC_NOT_SUPPORTED",
    I18n.t("authentication.cie.card.error.unknownCardContent")
  ],
  [
    "ON_TAG_DISCOVERED_NOT_CIE",
    I18n.t("authentication.cie.card.error.unknownCardContent")
  ],
  ["PIN Locked", I18n.t("authentication.cie.card.error.generic")],
  ["ON_CARD_PIN_LOCKED", I18n.t("authentication.cie.card.error.generic")],
  ["ON_PIN_ERROR", I18n.t("authentication.cie.card.error.tryAgain")],
  ["PIN_INPUT_ERROR", ""],
  ["CERTIFICATE_EXPIRED", I18n.t("authentication.cie.card.error.generic")],
  ["CERTIFICATE_REVOKED", I18n.t("authentication.cie.card.error.generic")],
  ["AUTHENTICATION_ERROR", I18n.t("authentication.cie.card.error.generic")],
  [
    "EXTENDED_APDU_NOT_SUPPORTED",
    I18n.t("authentication.cie.nfc.apduNotSupported")
  ],
  [
    "ON_NO_INTERNET_CONNECTION",
    I18n.t("authentication.cie.card.error.tryAgain")
  ],
  ["STOP_NFC_ERROR", ""],
  ["START_NFC_ERROR", ""]
]);

export type TextForState = {
  title: string;
  subtitle?: string;
  content: string;
};

// some texts changes depending on current running Platform
export const getTextForState = (
  state: ReadingState.waiting_card | ReadingState.error,
  errorMessage: string = ""
): TextForState => {
  const texts: Record<
    ReadingState.waiting_card | ReadingState.error,
    TextForState
  > = Platform.select({
    ios: {
      [ReadingState.waiting_card]: {
        title: I18n.t("authentication.cie.card.titleiOS"),
        subtitle: I18n.t("authentication.cie.card.layCardMessageHeaderiOS"),
        // the native alert hides the screen content and shows a message it self
        content: ""
      },
      [ReadingState.error]: {
        title: I18n.t("authentication.cie.card.error.readerCardLostTitle"),
        subtitle: "",
        // the native alert hides the screen content and shows a message it self
        content: ""
      },
      [ReadingState.reading]: {
        title: I18n.t("authentication.cie.card.titleiOS"),
        subtitle: I18n.t("authentication.cie.card.layCardMessageHeaderiOS"),
        // the native alert hides the screen content and shows a message it self
        content: ""
      }
    },
    default: {
      [ReadingState.waiting_card]: {
        title: I18n.t("authentication.cie.card.title"),
        subtitle: I18n.t("authentication.cie.card.layCardMessageHeader"),
        content: I18n.t("authentication.cie.card.layCardMessageFooter")
      },
      [ReadingState.error]: {
        title: I18n.t("authentication.cie.card.error.readerCardLostTitle"),
        subtitle: I18n.t("authentication.cie.card.error.onTagLost"),
        content: errorMessage
      }
    }
  });
  return texts[state];
};
