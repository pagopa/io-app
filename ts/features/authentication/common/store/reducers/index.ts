import { PersistPartial } from "redux-persist";
import { isActionOf } from "typesafe-actions";
import {
  clearCurrentSession,
  idpSelected,
  loginSuccess,
  logoutFailure,
  logoutSuccess,
  resetAuthenticationState,
  sessionExpired,
  sessionInformationLoadSuccess,
  sessionInvalid,
  logoutRequest
} from "../actions";
import { Action } from "../../../../../store/actions/types";
import { refreshSessionToken } from "../../../fastLogin/store/actions/tokenRefreshActions";
import { AuthenticationState, LoggedOutWithoutIdp } from "../models";
import {
  isLoggedIn,
  isLoggedOutWithIdp,
  isSessionExpired
} from "../utils/guards";

// Here we mix the plain AuthenticationState with the keys added by redux-persist
type PersistedAuthenticationState = AuthenticationState & PersistPartial;

// Initially the user is logged out and hasn't selected an IDP
export const INITIAL_STATE: LoggedOutWithoutIdp = {
  kind: "LoggedOutWithoutIdp",
  reason: "NOT_LOGGED_IN"
};

const authenticationReducer = (
  state: AuthenticationState = INITIAL_STATE,
  action: Action
  // eslint-disable-next-line sonarjs/cognitive-complexity
): AuthenticationState => {
  if (isActionOf(idpSelected, action) && !isLoggedIn(state)) {
    // Save the selected IDP in the state
    return {
      ...state,
      ...{
        kind: "LoggedOutWithIdp",
        idp: action.payload
      }
    };
  }

  if (isActionOf(loginSuccess, action) && isLoggedOutWithIdp(state)) {
    // Save the SessionToken (got from the WebView redirect url) in the state
    return {
      kind: "LoggedInWithoutSessionInfo",
      idp: state.idp,
      sessionToken: action.payload.token
    };
  }

  if (isActionOf(refreshSessionToken.success, action) && isLoggedIn(state)) {
    // Save the new SessionToken in the state
    return {
      ...state,
      ...{
        sessionToken: action.payload
      }
    };
  }

  if (isActionOf(sessionInformationLoadSuccess, action) && isLoggedIn(state)) {
    // Save the session info in the state
    return {
      ...state,
      ...{
        kind: "LoggedInWithSessionInfo",
        sessionInfo: action.payload
      }
    };
  }

  if (isActionOf(logoutRequest, action) && isLoggedIn(state)) {
    return {
      ...state,
      ...{
        kind: "LogoutRequested",
        reason: "NOT_LOGGED_IN"
      }
    };
  }

  if (
    (isActionOf(sessionExpired, action) ||
      isActionOf(sessionInvalid, action) ||
      isActionOf(logoutSuccess, action) ||
      isActionOf(logoutFailure, action)) &&
    isLoggedIn(state)
  ) {
    return {
      kind: "LoggedOutWithIdp",
      idp: state.idp,
      reason: isActionOf(sessionExpired, action)
        ? "SESSION_EXPIRED"
        : "NOT_LOGGED_IN"
    };
  }

  if (isActionOf(clearCurrentSession, action)) {
    return INITIAL_STATE;
  }

  if (isActionOf(resetAuthenticationState, action) && isSessionExpired(state)) {
    return INITIAL_STATE;
  }

  return state;
};

export type { AuthenticationState, PersistedAuthenticationState };
export default authenticationReducer;
