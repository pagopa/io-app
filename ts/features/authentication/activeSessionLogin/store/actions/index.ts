import { ActionType, createStandardAction } from "typesafe-actions";
import { SpidIdp } from "../../../../../utils/idps";
import { SessionToken } from "../../../../../types/SessionToken";

export const setActiveSessionLoginLocalFlag = createStandardAction(
  "SET_ACTIVE_SESSION_LOGIN_LOCAL_FLAG"
)<boolean>();

export const setStartActiveSessionLogin = createStandardAction(
  "SET_START_ACTIVE_SESSION_LOGIN"
)();

export const setRetryActiveSessionLogin = createStandardAction(
  "SET_RETRY_ACTIVE_SESSION_LOGIN"
)();

export const setIdpSelectedActiveSessionLogin = createStandardAction(
  "SET_IDP_SELECTED_ACTIVE_SESSION_LOGIN"
)<SpidIdp>();

export const activeSessionLoginFailure = createStandardAction(
  "ACTIVE_SESSION_LOGIN_FAILURE"
)();

export const activeSessionLoginSuccess = createStandardAction(
  "ACTIVE_SESSION_LOGIN_SUCCESS"
)<SessionToken>();

export const setFastLoginOptSessionLogin = createStandardAction(
  "SET_FAST_LOGIN_OPT_IN_ACTIVE_SESSION_LOGIN"
)<boolean>();

export const consolidateActiveSessionLoginData = createStandardAction(
  "CONSOLIDATE_ACTIVE_SESSION_LOGIN_DATA"
)<{
  token: SessionToken;
  idp: SpidIdp;
  fastLoginOptIn: boolean;
}>();

export const setFinishedActiveSessionLoginFlow = createStandardAction(
  "SET_FINISHED_ACTIVE_SESSION_LOGIN_FLOW"
)();

export const setLoggedOutUserWithDifferentCF = createStandardAction(
  "SET_LOGGED_OUT_USER_WITH_DIFFERENT_CF"
)();

export type LoginInfoActions =
  | ActionType<typeof setActiveSessionLoginLocalFlag>
  | ActionType<typeof setStartActiveSessionLogin>
  | ActionType<typeof setRetryActiveSessionLogin>
  | ActionType<typeof setIdpSelectedActiveSessionLogin>
  | ActionType<typeof activeSessionLoginFailure>
  | ActionType<typeof activeSessionLoginSuccess>
  | ActionType<typeof consolidateActiveSessionLoginData>
  | ActionType<typeof setFastLoginOptSessionLogin>
  | ActionType<typeof setFinishedActiveSessionLoginFlow>
  | ActionType<typeof setLoggedOutUserWithDifferentCF>;
