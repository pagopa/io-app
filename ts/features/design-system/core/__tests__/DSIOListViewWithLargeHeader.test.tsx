import { createStore } from "redux";
import { applicationChangeState } from "../../../../store/actions/application";
import { appReducer } from "../../../../store/reducers";
import { GlobalState } from "../../../../store/reducers/types";
import { renderScreenWithNavigationStoreContext } from "../../../../utils/testWrapper";
import DESIGN_SYSTEM_ROUTES from "../../navigation/routes";
import { DSIOListViewWithLargeHeader } from "../DSIOListViewWithLargeHeader";

jest.mock("react-native", () => ({
  AccessibilityInfo: {
    setAccessibilityFocus: jest.fn()
  }
}));

jest.mock("@react-navigation/native-stack", () => ({
  createNativeStackNavigator: () => ({
    Navigator: ({ children }: any) => <>{children}</>,
    Screen: () => null
  })
}));

jest.mock("@react-navigation/stack", () => ({
  createStackNavigator: () => ({
    Navigator: ({ children }: any) => <>{children}</>,
    Screen: () => null
  })
}));

jest.mock("@react-navigation/bottom-tabs", () => ({
  createBottomTabNavigator: () => ({
    Navigator: ({ children }: any) => <>{children}</>,
    Screen: () => null
  })
}));

jest.mock("@react-navigation/native", () => ({
  useNavigation: () => ({
    setOptions: jest.fn
  }),
  useRoute: jest.fn,
  useFocusEffect: jest.fn,
  useIsFocused: jest.fn,
  NavigationContainer: ({ children }: any) => <>{children}</>,
  DarkTheme: {
    dark: true,
    colors: {
      primary: "#000",
      background: "#000",
      card: "#111",
      text: "#fff",
      border: "#222",
      notification: "#333"
    }
  },
  DefaultTheme: {
    dark: false,
    colors: {
      primary: "#fff",
      background: "#fff",
      card: "#eee",
      text: "#000",
      border: "#ccc",
      notification: "#ddd"
    }
  }
}));

const renderComponent = () => {
  const globalState = appReducer(undefined, applicationChangeState("active"));
  return renderScreenWithNavigationStoreContext<GlobalState>(
    () => <DSIOListViewWithLargeHeader />,
    DESIGN_SYSTEM_ROUTES.SCREENS.IOLISTVIEW_LARGE_HEADER.route,
    {},
    createStore(appReducer, globalState as any)
  );
};

describe("DSIOListViewWithLargeHeader", () => {
  it("should render list items correctly", () => {
    const { toJSON } = renderComponent();
    expect(toJSON()).toMatchSnapshot();
  });
});
