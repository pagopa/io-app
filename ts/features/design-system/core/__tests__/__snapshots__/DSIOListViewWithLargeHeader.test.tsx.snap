// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DSIOListViewWithLargeHeader should render list items correctly 1`] = `null`;
